google/cloud/forseti/scanner/audit/role_rules_engine.py

```diff
@@ -0,0 +1,266 @@
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def (dict): A dictionary containing rule definition
+                properties.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+        """
+        if 'name' not in rule_def:
+            raise audit_errors.InvalidRulesSchemaError(
+                'Lack of role_name in rule {}'.format(rule_index))
+        if 'role_name' not in rule_def:
+            raise audit_errors.InvalidRulesSchemaError(
+                'Lack of role_name in rule {}'.format(rule_index))
+        role_name = rule_def['role_name']
+        if 'permissions' not in rule_def:
+            raise audit_errors.InvalidRulesSchemaError(
+                'Lack of permissions in rule {}'.format(rule_index))
+        if 'resource' not in rule_def:
+            raise audit_errors.InvalidRulesSchemaError(
+                'Lack of resource in rule {}'.format(rule_index))
+        res = rule_def['resource']
+
+        rule = Rule(rule_index=rule_index,
+                    rule_name=rule_def['name'],
+                    permissions=rule_def['permissions'],
+                    res=res)
+
+        if role_name not in self.rules_map:
+            self.rules_map[role_name] = [rule]
+        else:
+            self.rules_map[role_name].append(rule)
+
+    def get_rule_by_role_name(self, role_name):
+        """Get the rule of a given role.
+
+        Args:
+            role_name (str): Name of a role.
+        Returns:
+            list: Rule list of the given role.
+        """
+        return self.rules_map.get(role_name, [])
+
+
+class Rule(object):
+    """Rule properties from the rule definition file. Also finds violations."""
+
+    def __init__(self, rule_index, rule_name, permissions, res):
+        """Initialize.
+
+        Args:
+            rule_index (int): The index of the rule.
+            rule_name (str): Name of the rule.
+            permissions (int): Expected permissions of the role.
+            res (dict): Parent resource of the role that should obey the rule.
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.permissions = permissions[:]
+        self.res_types = res[:]
+
+        for res_item in self.res_types:
+            if 'type' not in res_item:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Lack of resource:type in rule {}'.format(rule_index))
+            if 'resource_ids' not in res_item:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Lack of resource:resource_ids in rule {}'.format(
+                        rule_index))
+
+            if '*' in res_item['resource_ids']:
+                res_item = ['*']
```

Comment:
This line won't change the item in `self.res_types`, it will just update the temporary variable `res_item`, which is then immediately replaced.